test/integration/framework.go

```diff
@@ -109,24 +120,45 @@ func withConfigGetFreshApiserverAndClient(
 		options.SecureServingOptions.ServerCert.CertDirectory = certDir
 
 		if err := server.RunServer(options, stopCh); err != nil {
-			close(serverFailed)
-			t.Logf("Error in bringing up the server: %v", err)
+			for err != nil {
+				if strings.Contains(err.Error(), "bind: address already in use") {
+					t.Logf("Failed to start Controller on port %v, %v.", securePort, err)
+					securePort = rand.Intn(55000) + 10000
+					atomic.StoreUint32(&controllerPort, uint32(securePort))
+					secureAddr = fmt.Sprintf("https://localhost:%d", securePort)
+					options.SecureServingOptions.BindPort = securePort
+					t.Logf("Retrying on %v", securePort)
+					err = server.RunServer(options, stopCh)
+				} else {
+					close(serverFailed)
+					t.Logf("Error in bringing up the server: %v", err)
+				}
+			}
 		}
 	}()
 
-	if err := waitForApiserverUp(secureAddr, serverFailed); err != nil {
+	// controllerUrl is dynamically built in the anonymous function below.  It may change if the launch code
+	// above hits a port already in use.
+	var controllerUrl string
+
+	if err := waitForApiserverUp(serverFailed, func() string {
```

Comment:
If there is in fact a collision, what prevents this from succeeding with the wrong client before the port number is switched out?